refactor: Apktool Remastered

IgorEisberg opened the pull request:
Refactor and reorganization of Apktool, especially resource decoding. ARSC parsing and resource file decoding rewritten, obviously keeping edge-case obfuscation handling intact.
Now it's a lot more congruent with AOSP source code, particularly based on libs/androidfw and tools/aapt2.
The code is clean and much easier to follow.
Everything that can be perceived as confusing is documented to avoid breaking changes later on.
Most of the changes are on the back end and won't be noticeable to the average user.

User-end changes include:

* Support for 32-bit platforms is now over.

* Legacy aapt ("aapt1") is now resting in peace. It's about time. It has limited the ability to evolve Apktool due to many compatibility issues. aapt2 is a lot more robust.

* Supports decoding resource files not directly under `res`/`r`/`R` folders.

* `-api`/`--api-level` option is no longer available or needed. Apktool and baksmali take care of setting the correct API to use for decoding/building dex files, you don't need to worry about it anymore.

* Advanced commands no longer occupy short options and only get long options. This is done for extensibility's sake, non-commonly used options should not take up short options intended for core options. Also, short options are restricted to a single character, which enables us to specify multiple options togethar, e.g. `-v -f -o output` can now be written as `-vfo output`. Before this change, combining short options was disabled and partial matching was enabled instead (rarely seen in CLI tools).

* `efd`/`empty-framework-dir` command is replaced with `cf`/`clean-frameworks`. Both `cf`/`clean-frameworks` and `lf`/`list-frameworks` have been extended: the `-t`/`--frame-tag` option now lets you delete/list frameworks matching a specific tag, as well as `-a`/`--all` option to delete/list all frameworks regardless of tag (this also replaced the `-f` option for these commands). If neither options are specified, only untagged frameworks are matched.

* `packageInfo` in apktool.yml (along with `forcedPackageId` and `renamedManifestPackage`) is replaced with `resourcesInfo`, where all settings related to the transformation of resources.arsc are included: `packageId`, `packageName`, `sparseResources`, `compactEntries`. All future settings related to resources.arsc should be included here. Note: `packageName` alters the package name in the resource table only! To change the package name of the app itself, change it in AndroidManifest.xml. If no `packageName` is defined in `resourcesInfo`, then the resource table inherits the app's package name.

* Many `<item>` tags are now replaced with tags corresponding to their respective resource types, particularly for resource references, that's the standard in AOSP source code and is perfectly supported by aapt2. This used to be an issue with legacy aapt. Some `<item>` tags are still necessary when the value's format isn't directly supported by the resource type. This is somewhat common with `dimen`, where some values are not one of `dimension`, `float` or `fraction` formats. Example: plain `5` in the `integer` format has to appear as an `<item>` tag, along with added `format="integer"` attribute, otherwise it'll be re-encoded as a `float`. The `<item>` tag is also necessary for resource reference as the value for a bag type (attr, array, plurals, style), as the usual tag for these types only parses child nodes (not text in body), but specifying `format="reference"` is optional, since aapt2 alwats parses the `@` prefix as a reference.

* `dp` is used instead of `dip` as it's the Android development standard, and more congruent with the other units: `px`, `sp`, `pt`, `in`, `mm`. Note: Both `dp` and `dip` are recognized when building and are encoded identically.

* `match_parent` is preferred over `fill_parent` that has been deprecated since API 8 (Android 2.2). It's also the standard in Android development for over a decade. Note: Both `match_parent` and `fill_parent` are recognized when building and are encoded identically.

* Hex integer values are written in compact form (e.g. `0x20` rather than `0x00000020`) to keep it clean and readable, and match original source codes as much as possible. The only exception is for resource IDs in public.xml. Those remain in the `0xPPTTEEEE` format regardless of value.

* Colors values are written exactly as they were written in source code. This respects the color data type in ARSC and doesn't reformat it (since rebuilding will encode it with a different color data type). In other words, if the source code had a color in the `#RRGGBB` (RGB8), `#ARGB` (ARGB4) or `#RGB` (RGB4) format, then that's how it'll be decoded. Previously, all colors were forced to the `#AARRGGBB` (ARGB8) format in output, regardless of original encoding data.

iBotPeaches commented:
I guess I shouldn't have changed the CI - looks like conflicts there. I thought you were asking me to relax the CI rules lol.

iBotPeaches commented:
@HassanMirza01 - I don't mind you taking part, but we are working on some big improvements, dropping old things and it seems you are just adding unrelated things into this thread leading to some confusion. I'm trying to review some large changes and unrelated res modes and MIUI is not really related.

aapt1 is going away that is final. I want to remove MIUI hacky workarounds, but they are already in codebase and don't take much effort to keep it.

IgorEisberg commented:
> > P.S. don't know what Hassan is talking about. I never decode with dummies and never had to.
> 
> am talking about this #3851

Don't know what to tell you, I don't have any resource IDs shifting around, in fact decoding-building-decoding again MiuiSystemUI.apk I have literally 0 diffs between the res folders.
Anyway, this is off topic, we are not here to discuss issues you're having, this is a development PR.

iBotPeaches commented:
Started the doc split - https://github.com/iBotPeaches/apktool.org/pull/17 WIP, once I get that to a point. I'll glean this PR so I can see what changed param wise to update docs.

IgorEisberg commented:
> Started the doc split - [iBotPeaches/apktool.org#17](https://github.com/iBotPeaches/apktool.org/pull/17) WIP, once I get that to a point. I'll glean this PR so I can see what changed param wise to update docs.

Did you know that you can even define a string-array like this?
```
    <array name="animals" format="string">
        <item>@string/racoon</item>
        <item>Rhino</item>
        <item>Elephant</item>
    </array>
```

And even this? XD
```
    <bag type="array" name="animals" format="string">
        <item>@string/racoon</item>
        <item>Rhino</item>
        <item>Elephant</item>
    </bag>
```

IgorEisberg commented:
Added support for OVERLAYABLE and OVERLAYABLE_POLICY chunks. Don't know how common it is, but good to have instead of skipping them. Also added a test.
